### ISSUE TITLE
CA1062 forgets non-null status from ValidatedNotNull after call to function containing delegate

### ISSUE THREAD

**Issue by mdro-bk** [dotnet/roslyn-analyzers#2919]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.6 (Latest)

#### Diagnostic ID

[CA1062](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1062-validate-arguments-of-public-methods?view=vs-2019)

#### Visual Studio version

16.3.3 (latest)

#### Repro steps

In the following minimized repro, accessing `foo` in `Test()` after the call to `Bar()` causes a false positive of CA1062 despite `Bar()` not having any way to access `foo`:
```c#
using System;

namespace ReproCA1062
{
    public static class Repro
    {
        public static void Test(string foo)
        {
            NotNull(foo, nameof(foo));
            int x = foo.Length; // no warning
            Bar();
            x = foo.Length; // CA1062 on foo
        }

        public static void Bar()
        {
            Action<int> hmm = x => { };
            hmm(0);
        }

        public static void NotNull([ValidatedNotNull] object param, string paramName)
        {
            if (param == null)
            {
                throw new ArgumentNullException(paramName);
            }
        }
    }

    [System.AttributeUsage(AttributeTargets.Parameter, AllowMultiple = false, Inherited = true)]
    public sealed class ValidatedNotNullAttribute : Attribute { }
}
```

Note that the bug does not appear if I use a local null-check:
```c#
public static void Test(string foo)
{
    if (foo == null)
    {
        throw new ArgumentNullException(nameof(foo));
    }
    int x = foo.Length; // no warning
    Bar();
    x = foo.Length; // no warning
}
```

**Comment by RobinHood70** [comment]
I can confirm that this is still an issue in 2.9.8. I have a function that calls two different delegates via `myAction?.Invoke()`. Variables validated before the call trigger CA1062 after the call. If I comment out the Invokes, CA1062 works as expected.

Bizarrely, copying the checked variable to a second variable and checking _that_ variable works as expected.
```
Before:
	var captionParser = WikiTextParser.Parse(link.Text);
	this.ReplaceNodes(page, captionParser);
	link.Text = WikiTextVisitor.Raw(captionParser); // CA1062 fires here.
```

```
After:
	var link2 = link ?? throw new InvalidOperationException();
	var captionParser = WikiTextParser.Parse(link2.Text);
	this.ReplaceNodes(page, captionParser);
	link2.Text = WikiTextVisitor.Raw(captionParser); // CA1062 passes unless throw is removed.
```
